src/bases/D3MPoolBase.sol

```diff
@@ -0,0 +1,145 @@
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+interface TokenLike {
+    function approve(address, uint256) external returns (bool);
+    function transfer(address, uint256) external returns (bool);
+    function transferFrom(address, address, uint256) external returns (bool);
+    function balanceOf(address) external view returns (uint256);
+}
+
+interface DaiJoinLike {
+    function dai() external view returns (address);
+    function join(address, uint256) external;
+    function exit(address, uint256) external;
+}
+
+interface CanLike {
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface d3mHubLike {
+    function vat() external view returns (address);
+}
+
+interface D3MPlanLike {
+    function calcSupplies(uint256) external view returns (uint256, uint256);
+    function maxBar() external view returns (uint256);
+}
+
+abstract contract D3MPoolBase {
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "D3MPoolBase/not-authorized");
+        _;
+    }
+
+    TokenLike   public immutable asset; // Dai
+    DaiJoinLike public immutable daiJoin;
+    address     public immutable hub;
+    address     public immutable pool;
+
+    address     public           share;
+    address     public           plan; // How we calculate target debt
+    uint256     public           live = 1;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Hope(address indexed dst, address indexed usr);
+    event Nope(address indexed dst, address indexed usr);
+    event Cage();
+    // --- EIP-4626 Events ---
+    event Deposit(address indexed caller, address indexed owner, uint256 assets, uint256 shares);
+    event Withdraw(address indexed caller, address indexed receiver, address indexed owner, uint256 assets, uint256 shares);
+
+    constructor(address hub_, address daiJoin_, address pool_) internal {
+
+        pool = pool_;
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        daiJoin = DaiJoinLike(daiJoin_);
```

Comment:
I think we could just store `daiJoin` in the `Hub` and do the `vat.dai <=> erc20` conversions there (which seem to be missing now).

Reply:
discussed offline and liked the suggestion: https://github.com/makerdao/dss-direct-deposit/pull/31/commits/a6af7528bad745fccd7be1d460bc1f84591a3cb8